pdx-extract/app/main.py

```diff
@@ -0,0 +1,174 @@
+import time
+
+import requests
+
+from app import config
+from app import pyppeteer
+from app import sanitize
+from app import storage
+from app import urls
+from app.logger import logger
+
+
+def authenticate_current_session(term, unique_session_id, cookies):
+    """Make a POST request that will authenticate the user with this JSESSIONID
+    and uniqueSessionId and enable the sched_page GET request to return JSON
+
+    :param term: Term dictionary with code and description keys.
+    :param unique_session_id: Unique session id generated by the page which
+                              allows authentication.
+    :param cookies: Cookies of the previous requests.
+    """
+
+    payload = {
+        "dataType": "json",
+        "endDatepicker": "",
+        "startDatepicker": "",
+        "studyPath": "",
+        "studyPathText": "",
+        "term": term["code"],
+        "uniqueSessionId": unique_session_id,
+    }
+    return requests.post(
+        urls.SEARCH_URL,
+        headers={"referer": urls.INIT_URL},
+        cookies=cookies,
+        params=payload,
+    )
+
+
+def get_schedule_json(subject, term, unique_session_id, cookies):
+    """Gets JSON representation of the subject for the specified term.
+
+    :param subject: The subject in question.
+    :param term: The term in question.
+    :param unique_session_id: Unique session id generated by the page which
+                              allows authentication.
+    :param cookies: Cookies of the previous requests.
+    """
+
+    payload = {
+        "txt_subject": subject["code"],
+        "txt_term": term["code"],
+        "startDatepicker": "",
+        "endDatepicker": "",
+        "uniqueSessionId": unique_session_id,
+        "pageOffset": "0",
+        "pageMaxSize": "100",
+        "sortColumn": "subjectDescription",
+        "sortDirection": "asc",
+    }
+    res = requests.get(
+        urls.SCHEDULE_URL,
+        headers={"referer": urls.CLASS_URL},
+        cookies=cookies,
+        params=payload,
+    )
+
+    if res.ok:
+        return res.json()
+
+    return None
+
+
+def get_subjects(cookies, unique_session_id, term_date):
+    """Gets the subjects that are available for a particular term.
+
+    :param cookies: Cookies needed to authenticate the request.
+    :param unique_session_id: Parameter needed to authenticate the request.
+    :param term_date: term Where the subjects will be searched for.
+    :returns: JSON with list of subjects
+    """
+
+    payload = {
+        "uniqueSessionId": unique_session_id,
+        "dataType": "json",
+        "searchTerm": "",
+        "term": term_date,
+        "offset": "1",
+        "max": config.MAX_SUBJECTS,
+        # Query string params expect a timestamp with extra 3 digits
+        "_:": str(int(time.time() * 1000)),
+    }
+    res = requests.get(urls.SUBJECTS_URL, cookies=cookies, params=payload)
+
+    if res.ok:
+        return res.json()
```

Comment:
More meaningful to shortcut failure cases, so the return at the end of a function is the happy case:

```python
if not res.ok:
    return None

return res.json()
```

Reply:
I am confused about why there is not a `none` return type here, when the `get_schedule` function has one.

Reply:
Like great ideas, it's obvious in hindsight. Shortcutting the failure case and returning the positive case also documents the _intention_ of the function.